packages/server/test/preHook.test.ts

```diff
@@ -0,0 +1,167 @@
+      .preHook(preHook)
+      .query('foo1', {
+        resolve() {
+          return 'bar1';
+        },
+      })
+      .query('foo2', {
+        resolve() {
+          return 'bar2';
+        },
+      }),
+  );
+
+  expect(await client.query('foo1')).toBe('bar1');
+  expect(await client.query('foo2')).toBe('bar2');
+  expect(preHook).toHaveBeenCalledTimes(2);
+  close();
+});
+
+test('is not called if def last', async () => {
+  const preHook = jest.fn();
+  const { client, close } = routerToServerAndClient(
+    trpc
+      .router()
+      .query('foo', {
+        resolve() {
+          return 'bar';
+        },
+      })
+      .preHook(preHook),
+  );
+
+  expect(await client.query('foo')).toBe('bar');
+  expect(preHook).toHaveBeenCalledTimes(0);
+  close();
+});
+
+test('allows you to throw an error (e.g. auth)', async () => {
+  type Context = {
+    user?: {
+      id: number;
+      name: string;
+      isAdmin: boolean;
+    };
+  };
+  const resolverMock = jest.fn();
+
+  const headers: Record<string, string | undefined> = {};
+
+  const { client, close } = routerToServerAndClient(
+    trpc
+      .router<Context>()
+      .query('foo', {
+        resolve() {
+          return 'bar';
+        },
+      })
+      .merge(
+        'admin.',
+        trpc
+          .router<Context>()
+          .preHook(({ ctx }) => {
+            if (!ctx.user?.isAdmin) {
+              throw httpError.unauthorized();
+            }
+          })
+          .query('secretPlace', {
+            resolve() {
+              resolverMock();
+
+              return 'a key';
+            },
+          }),
+      ),
+    {
+      createContext({ req }) {
+        if (req.headers.authorization === 'meow') {
+          return {
+            user: {
+              id: 1,
+              name: 'KATT',
+              isAdmin: true,
+            },
+          };
+        }
+        return {};
+      },
+      getHeaders: () => headers,
+    },
+  );
+
+  expect(await client.query('foo')).toBe('bar');
+  await expect(client.query('admin.secretPlace')).rejects.toMatchObject({
+    res: { status: 401 },
+  });
+  expect(resolverMock).toHaveBeenCalledTimes(0);
+  headers.authorization = 'meow';
+  expect(await client.query('admin.secretPlace')).toBe('a key');
+  expect(resolverMock).toHaveBeenCalledTimes(1);
+  close();
+});
+
+test('child routers + hook call order', async () => {
+  const preHookInParent = jest.fn();
+  const preHookInChild = jest.fn();
+  const preHookInGrandChild = jest.fn();
+  const { client, close } = routerToServerAndClient(
+    trpc
+      .router()
+      .preHook(preHookInParent)
+      .query('name', {
+        resolve() {
+          return 'GrandPa';
+        },
+      })
+      .merge(
+        'child.',
+        trpc
+          .router()
+          .preHook(preHookInChild)
+          .query('name', {
+            resolve() {
+              return 'Parent';
+            },
+          })
+          .merge(
+            'child.',
```

Comment:
I dont get it. Can we merge different routers and specify the same name (`child.`) twice?

Reply:
nevermind. I see lower down. so calling it `.child` actually appends to the previous route. That is actually good to know. It means I cannot do 

`mainRouter.merge('products.',productsRouter).merge('users.', usersRouter) as it will mean to query on users I'd need to go `products.users.list` which is fundamentally wrong as users is not a subroute to products, right?

Reply:
you don't _have to_ use a prefix but you can if you want

Reply:
these two examples are equivalent:

```ts

  type Context = {
    user?: {
      id: number;
      name: string;
      isAdmin: boolean;
    };
  };
  // **using prefix**
  trpc
    .router()
    .query('foo', {
      resolve() {
        return 'bar';
      },
    })
    .merge(
      'admin.',
      trpc
        .router<Context>()
        .preHook(({ ctx }) => {
          if (!ctx.user?.isAdmin) {
            throw httpError.unauthorized();
          }
        })
        .query('secretPlace', {
          resolve() {
            return 'a key';
          },
        }),
    );

  // **not** using prefix
  trpc
    .router()
    .query('foo', {
      resolve() {
        return 'bar';
      },
    })
    .merge(
      trpc
        .router<Context>()
        .preHook(({ ctx }) => {
          if (!ctx.user?.isAdmin) {
            throw httpError.unauthorized();
          }
        })
        .query('admin.secretPlace', {
          resolve() {
            return 'a key';
          },
        }),
    );
```

Reply:
In my app I've split up my routers in different files

<details><summary>api/trpc/__shared__.tsx</summary>

```ts
import { CreateExpressContextOptions } from '@trpc/server/dist/adapters/express';
import * as trpc from '@trpc/server';
import { inferAsyncReturnType } from '@trpc/server';
import { createUserSdk, decodeAuthToken } from '../__shared__';

export const createContext = async ({ req }: CreateExpressContextOptions) => {
  const auth0 = await decodeAuthToken(req);
  const roles = auth0['https://hasura.io/jwt/claims']['x-hasura-allowed-roles'];
  const sdk = createUserSdk(auth0);

  return {
    sdk,
    user: {
      auth0,
      roles,
    },
  };
};
export type Context = inferAsyncReturnType<typeof createContext>;
export function createRouter() {
  return trpc.router<Context>();
}
```
</details>

<details><summary>api/trpc/[...trpc].ts</summary>

```ts
import * as z from 'zod';
import { adminRouter } from './admin';
import { createRouter, createContext } from './__shared__';
import { createExpressMiddleware } from '@trpc/server/dist/adapters/express';
export type { Context } from './__shared__';

const appRouter = createRouter()
  .query('hello', {
    input: z
      .object({
        text: z.string().optional(),
      })
      .optional(),
    resolve({ input }) {
      return {
        greeting: `hello ${input?.text ?? 'world'}`,
      };
    },
  })
  .merge('admin.', adminRouter);

// Exporting type _type_ AppRouter only exposes types that can be used for inference
// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export
export type AppRouter = typeof appRouter;

export const trpcMiddleware = createExpressMiddleware({
  router: appRouter,
  createContext,
});

```
</details>


<details><summary>api/trpc/admin.ts</summary>

```ts
import { createRouter } from './__shared__';
import * as z from 'zod';
import { httpError } from '@trpc/server';
import gql from 'graphql-tag';
import { v4 } from 'uuid';
import { UserRole_Enum } from '../__generated__/sdk';
import fetch from 'node-fetch';
import { env } from '../__shared__';

gql`
  mutation admin__addUser(
    $user: User_insert_input!
    $organization: Organization_insert_input!
    $userId: uuid!
    $organizationId: uuid!
  ) {
    insert_Organization_one(object: $organization) {
      id
    }
    insert_User_one(object: $user) {
      id
    }
    insert_OrganizationUser_one(
      object: { userId: $userId, organizationId: $organizationId }
    ) {
      userId
    }
  }
`;
function randomString(len: number) {
  const charSet =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  var randomString = '';
  for (let i = 0; i < len; i++) {
    const randomPoz = Math.floor(Math.random() * charSet.length);
    randomString += charSet.substring(randomPoz, randomPoz + 1);
  }
  return randomString;
}

export const adminRouter = createRouter().mutation('createUser', {
  input: z.object({
    user: z.object({
      name: z.string().min(1),
      email: z.string().email(),
    }),
    organization: z.object({
      name: z.string(),
      slug: z.string().regex(/^[a-z0-9-]+$/),
      url: z.string().optional(),
      businessType: z.string(),
      country: z.string(),
    }),
  }),
  async resolve({ ctx, input }) {
    if (!ctx.user.roles.includes('admin')) {
      throw httpError.unauthorized();
    }
    const userId = v4();
    const organizationId = v4();

    await ctx.sdk.admin__addUser({
      userId,
      organizationId,
      organization: {
        id: organizationId,
        ...input.organization,
      },
      user: {
        id: userId,
        ...input.user,
        role: UserRole_Enum.User,
      },
    });
    const headers = await (
      await fetch(`https://${env.AUTH0_MGMT_DOMAIN}/oauth/token`, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
        },
        body: JSON.stringify({
          client_id: env.AUTH0_MGMT_CLIENT_ID,
          client_secret: env.AUTH0_MGMT_CLIENT_SECRET,
          audience: `https://${env.AUTH0_MGMT_DOMAIN}/api/v2/`,
          grant_type: 'client_credentials',
        }),
      })
    ).json();
    const token = headers.access_token;
    if (typeof token !== 'string') {
      throw new Error('Coudl not get token');
    }
    const createUserResponse = await (
      await fetch(`https://${env.AUTH0_MGMT_DOMAIN}/api/v2/users`, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({
          email: input.user.email,
          name: input.user.name,
          connection: 'Username-Password-Authentication',
          password: randomString(32),
        }),
      })
    ).json();

    if (
      !createUserResponse.created_at ||
      createUserResponse.email !== input.user.email
    ) {
      console.error('Auth0 response', createUserResponse);
      throw new Error(
        'User not created - auth0 said something unexpected - does the user already exist maybe?',
      );
    }

    return null;
  },
});

```
</details>

(kinda using hasura as an sdk on the backend which isn't amazing but it gives me typesafety db access and I'm in process of moving to prisma)

Reply:
yep makes sense. Most of our stuff is on prisma finally. Took months to get the team to adopt it